### PR TITLE
[SPARK-36669][SQL] Add Lz4 wrappers for Hadoop Lz4 codec

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to add a few LZ4 wrapper classes for Parquet Lz4 compression output that uses Hadoop Lz4 codec.

### Why are the changes needed?

Currently we use Hadop 3.3.1's shaded client libraries. Lz4 is a provided dependency in Hadoop Common 3.3.1 for Lz4Codec. But it isn't excluded from relocation in these libraries. So to use lz4 as Parquet codec, we will hit the exception even we include lz4 as dependency.

```
[info]   Cause: java.lang.NoClassDefFoundError: org/apache/hadoop/shaded/net/jpountz/lz4/LZ4Factory                                                                                            
[info]   at org.apache.hadoop.io.compress.lz4.Lz4Compressor.<init>(Lz4Compressor.java:66)
[info]   at org.apache.hadoop.io.compress.Lz4Codec.createCompressor(Lz4Codec.java:119)                                                                                                         
[info]   at org.apache.hadoop.io.compress.CodecPool.getCompressor(CodecPool.java:152)                                                                                                          
[info]   at org.apache.hadoop.io.compress.CodecPool.getCompressor(CodecPool.java:168)     
```

Before the issue is fixed at Hadoop new release, we can add a few wrapper classes for Lz4 codec.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Modified test.

### CODE REVIEW

**Comment by viirya** [comment]
After discussing with some PMC members, more preferred direction now is to document snappy issue as known issue for 3.2. Here to deal with lz4 issue, so to add lz4 wrappers back.

cc @dbtsai @dongjoon-hyun @sunchao @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala [hunk 1]
```diff
@@ -56,13 +56,12 @@ class ParquetCodecSuite extends FileSourceCodecSuite {
   override def format: String = "parquet"
   override val codecConfigName: String = SQLConf.PARQUET_COMPRESSION.key
   // Exclude "lzo" because it is GPL-licenced so not included in Hadoop.
-  // TODO(SPARK-36669): "lz4" codec fails due to HADOOP-17891.
   override protected def availableCodecs: Seq[String] =
     if (System.getProperty("os.arch") == "aarch64") {
       // Exclude "brotli" due to PARQUET-1975.
-      Seq("none", "uncompressed", "snappy", "gzip", "zstd")
+      Seq("none", "uncompressed", "snappy", "lz4", "gzip", "zstd")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:62]: Sorry I may miss some context. What's the problem with snappy? It's tested here.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceCodecSuite.scala:62]: Parquet uses its snappy codec, not hadoop's. So it is not affected.

**Review by cloud-fan — APPROVED** [review]

**Comment by viirya** [comment]
Thanks! Merging to master/3.2.

**Comment by dongjoon-hyun** [comment]
+1, late LGTM like the first PR. Thank you, @viirya and @cloud-fan .

**Comment by gengliangwang** [comment]
Late +1, thanks for the fix!

**Pull request closed without merging**